File: backend/app/schemas/claim.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.utils.uploads import public_upload_name



class ClaimResponse(BaseModel):
    id: int
    claim_date: date
    total_km: float
    travel_total: float
    daily_allowance: float
    grand_total: float
    patient_visited_today: bool | None = None
    status: str
    therapist_name: str | None = None
    patient_count: int | None = 0
    per_km_rate: float | None = 0
    rejection_reason: str | None = None
    reviewed_at: datetime | None = None
    reviewed_by: int | None = None
    revision: int = 1
    policy_id: int | None = None
    calculation_version: str = "decimal-v1"
    rounding_mode: str = "ROUND_HALF_UP"
    included_travel_ids: list[int] | None = None

    model_config = ConfigDict(from_attributes=True)


class ClaimDetailsClaimResponse(BaseModel):
    id: int
    therapist_id: int
    therapist_name: str | None = None
    therapist_role: str | None = None
    claim_date: date
    submitted_at: datetime | None = None
    total_km: float
    per_km_rate: float | None = 0
    travel_total: float
    daily_allowance: float
    grand_total: float
    status: str
    notes: str | None = None
    patient_count: int = 0
    rejection_reason: str | None = None
    reviewed_at: datetime | None = None
    reviewed_by: int | None = None
    revision: int = 1


class ClaimTravelEntryResponse(BaseModel):
    id: int
    travel_date: date
    travel_timestamp: datetime | None = None
    patient_name: str | None = None
    transport_mode: str
    bill_amount: float | None = None
    invoice_file: str | None = None
    from_address: str
    to_address: str
    total_km: float
    per_km_rate: float
    travel_fare: float
    patient_visited: bool
    status: str

    @field_validator("invoice_file", mode="before")
    @classmethod
    def serialize_invoice_name(cls, value) -> str | None:
        return public_upload_name(value, "invoice")


class ClaimDetailsResponse(BaseModel):
    claim: ClaimDetailsClaimResponse
    travels: list[ClaimTravelEntryResponse]


class ClaimRejectRequest(BaseModel):
    rejection_reason: str = Field(min_length=1, max_length=500)

    @field_validator("rejection_reason")
    @classmethod
    def validate_rejection_reason(cls, value: str) -> str:
        reason = value.strip()
        if not reason:
            raise ValueError("Rejection reason is required")
        return reason

File: mobile/src/types/claim.ts
export interface ClaimResponse {
  id: number;
  claim_date: string;
  total_km: number;
  travel_total: number;
  daily_allowance: number;
  grand_total: number;
  patient_visited_today?: boolean | null;
  status: string;
  therapist_name?: string | null;
  patient_count?: number | null;
  per_km_rate?: number | null;
  rejection_reason?: string | null;
  reviewed_at?: string | null;
  reviewed_by?: number | null;
  revision?: number;
}

export interface ClaimTravelEntry {
  id: number;
  travel_date: string;
  travel_timestamp?: string | null;
  patient_name: string | null;
  transport_mode: string;
  bill_amount: number | null;
  invoice_file: string | null;
  from_address: string;
  to_address: string;
  total_km: number;
  per_km_rate: number;
  travel_fare: number;
  patient_visited: boolean;
  status: string;
}

export interface ClaimDetailsResponse {
  claim: ClaimResponse & {
    therapist_id: number;
    therapist_role?: string | null;
    submitted_at?: string | null;
    notes?: string | null;
  };
  travels: ClaimTravelEntry[];
}

export interface ClaimReadinessBlocker {
  code: string;
  message: string;
  recoverable: boolean;
  suggested_action: string | null;
  affected_count: number;
  blocking_fields: string[];
}

export interface ClaimReadinessBase {
  business_date: string;
  state: "ready" | "blocked" | "already_submitted";
  can_submit: boolean;
  submission_mode: "submit" | "resubmit" | null;
  eligible_record_count: number;
  eligible_record_ids: number[];
  pending_review_count: number;
  existing_claim_id: number | null;
  existing_claim_status: string | null;
  existing_claim_revision: number | null;
  rejection_reason: string | null;
  total_amount: number;
  total_source: "eligible_records" | "existing_claim" | "none";
  calculation_version: string;
  rounding_mode: string;
  available_actions: string[];
  blocking_reasons: ClaimReadinessBlocker[];
  next_action: string | null;
}

export interface TherapistClaimReadiness extends ClaimReadinessBase {
  total_km: number;
  per_km_rate: number | null;
  travel_total: number;
  daily_allowance: number;
  patient_visited_today: boolean;
  policy_id: number | null;
  policy_version: number | null;
  policy_effective_from: string | null;
}

export interface DoctorClaimReadiness extends ClaimReadinessBase {
  expense_total: number;
}
